File: src/logic/pieces/pawn.py
from logic.move import Move
from logic.pieces.bishop import Bishop
from logic.pieces.knight import Knight
from logic.pieces.piece import Colour, Piece
from logic.pieces.queen import Queen
from logic.pieces.rook import Rook
from logic.position import Position

class Pawn(Piece):
    def legal_moves(self, board, / , looking_for_check = False) -> list[Move]:
        ret = []

        # can we capture to the left?
        if self.pos.x > 0 and (
            (self.colour == Colour.WHITE and (capturable_piece := board.piece_at(self.pos.x - 1, self.pos.y + 1)))
            or
            (self.colour == Colour.BLACK and (capturable_piece := board.piece_at(self.pos.x - 1, self.pos.y - 1)))
        ):
            if capturable_piece.colour != self.colour:
                if (self.colour == Colour.WHITE and capturable_piece.pos.y == 7) or (self.colour == Colour.BLACK and capturable_piece.pos.y == 0):
                    for piece in [Queen, Knight, Bishop, Rook]:
                        ret.append(Move(self, capturable_piece.pos, is_capturing=True, promotes_to=piece))
                else:
                    ret.append(Move(self, capturable_piece.pos, is_capturing = True))

        # can we capture to the right?
        if self.pos.x < 7 and (
            (self.colour == Colour.WHITE and (capturable_piece := board.piece_at(self.pos.x + 1, self.pos.y + 1)))
            or
            (self.colour == Colour.BLACK and (capturable_piece := board.piece_at(self.pos.x + 1, self.pos.y - 1)))
        ):
            if capturable_piece.colour != self.colour:
                if (self.colour == Colour.WHITE and capturable_piece.pos.y == 7) or (self.colour == Colour.BLACK and capturable_piece.pos.y == 0):
                    for piece in [Queen, Knight, Bishop, Rook]:
                        ret.append(Move(self, capturable_piece.pos, is_capturing=True, promotes_to=piece))
                else:
                    ret.append(Move(self, capturable_piece.pos, is_capturing = True))

        # -- Can we capture en passant?
        if board._en_passant_target is not None and \
            board._en_passant_target.pos.y == self.pos.y and (
                board._en_passant_target.pos.x == self.pos.x - 1
                or board._en_passant_target.pos.x == self.pos.x + 1
        ):
            if board._en_passant_target.colour != self.colour:
                old_pos = board._en_passant_target.pos
                new_pos = Position(old_pos.x, old_pos.y + (1 if self.colour == Colour.WHITE else -1))
                ret.append(Move(self, new_pos, is_capturing = True, en_passant = True))

        # -- Normal moves
        if self.colour == Colour.WHITE:
            for dy in range(1, 3 if self.pos.y == 1 else 2):
                y = self.pos.y + dy
                if y > 7 or board.piece_at(self.pos.x, y):
                    break
                pos = Position(self.pos.x, y)
                if y == 7:
                    for piece in [Queen, Knight, Bishop, Rook]:
                        ret.append(Move(self, pos, promotes_to=piece))
                else:
                    ret.append(Move(self, pos, becomes_en_passant_target=dy==2))
        else:
            for dy in range(1, 3 if self.pos.y == 6 else 2):
                y = self.pos.y - dy
                if y < 0 or board.piece_at(self.pos.x, y):
                    break
                pos = Position(self.pos.x, y)
                if y == 0:
                    for piece in [Queen, Knight, Bishop, Rook]:
                        ret.append(Move(self, pos, promotes_to=piece))
                else:
                    ret.append(Move(self, pos, becomes_en_passant_target=dy==2))

        if not looking_for_check and board.is_check_for(self.colour):
            return self.keep_only_blocking(ret, board)

        return ret

File: cpp/src/pieces/pawn.cpp
#include "../board.hpp"
#include "../coords.hpp"
#include "../move.hpp"
#include "piece.hpp"

std::vector<Move> pawn_moves(const Board& b, const Coords xy) {
    std::vector<Move> ret{};
    int8_t my_colour = b.colour_at(xy);

    // -- Capture to the left
    if (xy.x > 0) {
        int dy = my_colour == Colour::White ? 1 : -1;
        Coords left{xy.x - 1, xy.y + dy};
        int8_t capturable_piece = b.squares[left.to_index()];
        if (capturable_piece != 0) {
            if (my_colour != b.colour_at(left))
                if ((my_colour == White && left.y == 7)
                    || (my_colour == Black && left.y == 0))

                    for (auto piece : {Rook, Knigt, Bishop, Queen})
                        ret.push_back(Move{
                            xy.to_index(),
                            left.to_index(),
                            .is_capturing = true,
                            .promoting_to = (int8_t) (my_colour | piece)
                        });
                else
                    ret.push_back(Move{xy.to_index(), left.to_index()});
        }
    }

    // -- Capture to the right
    if (xy.x < 7) {
        int dy = my_colour == Colour::White ? 1 : -1;
        Coords right{xy.x + 1, xy.y + dy};
        int8_t capturable_piece = b.squares[right.to_index()];
        if (capturable_piece != 0) {
            if (my_colour != b.colour_at(right))
                if ((my_colour == White && right.y == 7)
                    || (my_colour == Black && right.y == 0))

                    for (auto piece : {Rook, Knigt, Bishop, Queen})
                        ret.push_back(Move{
                            xy.to_index(),
                            right.to_index(),
                            .is_capturing = true,
                            .promoting_to = (int8_t) (my_colour | piece)
                        });
                else
                    ret.push_back(Move{xy.to_index(), right.to_index()});
        }
    }

    // -- Capture en passant
    if (b.en_passant_target != -1) {
        Coords c = Coords::from_index(b.en_passant_target);
        int dy = my_colour == Colour::White ? 1 : -1;
        if (c.y == xy.y + dy && (c.x == xy.x - 1 || c.x == xy.x + 1)) {
            ret.push_back(Move{
                xy.to_index(),
                c.to_index(),
                .is_capturing = true,
                .en_passant = true
            });
        }
    }

    // -- Normal move + promotion
    bool is_on_starting_rank =
        my_colour == Colour::White ? xy.y == 1 : xy.y == 6;
    int max_dy = is_on_starting_rank ? 3 : 2;
    for (int dy = 1; dy < max_dy; dy++) {
        int actual_dy = my_colour == Colour::White ? dy : -dy;
        Coords new_xy{xy.x, xy.y + actual_dy};
        if (b.squares[new_xy.to_index()] != Piece::None)
            break;
        if (new_xy.y == 7 || new_xy.y == 0)
            for (auto piece : {Rook, Knigt, Bishop, Queen})
                ret.push_back(Move{
                    xy.to_index(),
                    new_xy.to_index(),
                    .promoting_to = (int8_t) (my_colour | piece)
                });
        else
            ret.push_back(Move{
                xy.to_index(),
                new_xy.to_index(),
            });
    }

    return ret;
}
